075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,435 @@
+2. From the **2nd** to the **(N-1)th** reconciliation (rebalancing is going on, checks when it's done):
+   * the `podSet()` does nothing because no new brokers are added this time.
+   * the `KafkaAutorebalancingReconciler().reconcile()` has to check the status of the rebalancing by looking at the corresponding `KafkaRebalance` custom resource status which will move through `ProposalPending`, `ProposalReady` to `Rebalancing` (with auto-approval). The rebalancing status is reflected into the `Kafka.status.autoRebalance.state` field leaving the `RebalancingOnScaleUp` value.
+3. On the **Nth** reconciliation (rebalancing is done):
+   * the `podSet()` does nothing because no new brokers are added this time.
+   * the `KafkaAutorebalancingReconciler().reconcile()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed). The end of rebalancing is reflected into the `Kafka.status.autoRebalance.state` field and the `KafkaRebalance` is deleted.
+
+Because the rebalancing runs across several reconciliations, different scenarios could happen even taking into account operator restarts:
+
+* the `KafkaCluster.addedNodes()` list is empty:
+  * the `Kafka.status.autoRebalance.state` is at `RebalancingOnScaleUp`. It's the normal flow, so just checking the rebalancing state.
+  * the `Kafka.status.autoRebalance.state` doesn't exist or it's `Ready`. It means no scale up requested. Nothing to do in terms of rebalancing.
+* the `KafkaCluster.addNodes()` list is not empty:
+  * the `Kafka.status.autoRebalance.state` doesn't exist or it's `Ready`. It's the normal flow, a new rebalance has to start.
+  * the `Kafka.status.autoRebalance.state` is at `RebalancingOnScaleUp`. The newly added brokers' IDs are compared with the one in the `Kafka.status.autoRebalance.modes.brokers` field. If they are equal, nothing to do (even if this situation should not be possible). If they are different, the user could have asked for another scale up right after a previous one. The operator has to stop the current rebalancing, by applying the `strimzi.io/rebalance: stop` annotation on the corresponding `KafkaRebalance` custom resource, and start a new rebalancing, by applying the `strimzi.io/rebalance: refresh` annotation, with a consistent list of brokers defined by the multiple requested scales up.
```

Comment:
Will this be synchronous or asynchronous? I assume it has to be async because the KafkaRebalanceAssmeblyOperator now also works in 2 minute beat for synchronous would block the rebalance cancelation for a long time. But if it is async, how will it work?

Reply:
Yeah, stopping and refreshing adds complexity but I just double checked, we support `refresh` during `PendingProposal` and `Rebalancing` states so no need to stop, it's done automatically.
So we can change here about using the `strimzi.io/rebalance: refresh` annotation directly which triggers the stop and a new request in one reconciliation (which starts immediately because of the rebalance custom resource change).